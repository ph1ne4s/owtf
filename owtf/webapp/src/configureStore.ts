/**
 * Create the store with dynamic reducers
 */

import { createStore, applyMiddleware, compose } from "redux";
import { fromJS } from "immutable";
import { routerMiddleware } from "react-router-redux";
import createSagaMiddleware from "redux-saga";
import { createLogger } from "redux-logger";
import createReducer from "./reducers";
import rootSaga from "./sagas";

const sagaMiddleware = createSagaMiddleware();

export default function configureStore(initialState = {}, history) {
  // Create the store with two middlewares
  // 1. sagaMiddleware: Makes redux-sagas work
  // 2. routerMiddleware: Syncs the location/URL path to the state
  const middlewares = [sagaMiddleware, routerMiddleware(history)];

  // Add loggerMiddleware if using app in development mode.
  if (process.env.NODE_ENV !== "production") {
    const logger = createLogger();
    middlewares.push(logger);
  }

  const enhancers = [applyMiddleware(...middlewares)];
  /**
   * If Redux DevTools Extension is installed use it, otherwise use Redux compose
   * http://extension.remotedev.io/#usage
   * https://medium.com/@zalmoxis/improve-your-development-workflow-with-redux-devtools-extension-f0379227ff83
   */
  const composeEnhancers =
    process.env.NODE_ENV !== "production" &&
    typeof window === "object" &&
    //@ts-ignore
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
      ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({
          // TODO Try to remove when `react-router-redux` is out of beta, LOCATION_CHANGE should not be fired more than once after hot reloading
          shouldHotReload: false
        })
      : compose;

  const store = createStore(
    createReducer(),
    fromJS(initialState),
    composeEnhancers(...enhancers)
  );

  sagaMiddleware.run(rootSaga);

  // Make reducers hot reloadable.
  if (module.hot) {
    module.hot.accept("./reducers", () => {
      const nextRootReducer = require("./reducers").default;
      store.replaceReducer(nextRootReducer);
    });
  }

  return store;
}
