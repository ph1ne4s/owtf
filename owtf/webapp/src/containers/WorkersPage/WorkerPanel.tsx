/* Worker Panel component
 *
 * Renders worker details individually inside a panel component
 *
 */
import React from "react";
import Moment from "react-moment";
import { Link } from "react-router-dom";

import { GrFormClose } from "react-icons/gr";
import { GrPauseFill } from "react-icons/gr";
import { GrPlayFill } from "react-icons/gr";

interface propsType {
  worker: any;
  resumeWorker: (worker_id: any) => void;
  pauseWorker: (worker_id: any) => void;
  abortWorker: (worker_id: any) => void;
  deleteWorker: (worker_id: any) => void;
  handleLogDialogShow: () => void;
  handleLogDialogContent: (logs: any) => void;
  workerLogs: any;
  onFetchWorkerLogs: Function;
  logDialogContent: string;
  openDialog: Function;
  key: any;
}
interface stateType {
  showLogs: boolean;
}

const panelStyle = {
  primary: "#337ab7",
  info: "#bce8f1",
  default: "#ddd"
};

export default class WorkerPanel extends React.Component<propsType, stateType> {
  constructor(props) {
    super(props);

    this.getPanelStyle = this.getPanelStyle.bind(this);
    this.getControlButtons = this.getControlButtons.bind(this);
    this.getWorkerLog = this.getWorkerLog.bind(this);
    this.displayLog = this.displayLog.bind(this);
    this.hideLog = this.hideLog.bind(this);
    this.openLogModal = this.openLogModal.bind(this);

    this.state = {
      showLogs: false
    };
  }

  /**
   * Function handles the state of Show log button [BUTTON/MENU]
   * Renders the show log menu
   */
  displayLog() {
    this.setState((state, props) => ({
      showLogs: !state.showLogs
    }));

    this.getWorkerLog(this.props.worker.name, -1);
  }

  /**
   * Function handles the state of Show log button [BUTTON/MENU]
   * Renders the show log button
   */
  hideLog() {
    this.setState((state, props) => ({
      showLogs: !state.showLogs
    }));
  }

  /**
   * Function handles the background state of worker panel
   */
  getPanelStyle() {
    const worker = this.props.worker;
    if (worker.busy && !worker.paused) {
      return panelStyle.primary;
    } else if (worker.paused) {
      return panelStyle.info;
    } else {
      return panelStyle.default;
    }
  }

  /**
   * Function to get control buttons based on the present state of a worker
   * It return pause button if worker is active & vice versa
   */
  getControlButtons() {
    const worker = this.props.worker;
    if (worker.busy) {
      if (worker.paused) {
        return (
          <button onClick={() => this.props.resumeWorker(worker.id)}>
            <GrPlayFill />
          </button>
        );
      } else {
        return (
          <div style={{ display: "flex", margin: "0.5rem" }}>
            <button onClick={() => this.props.pauseWorker(worker.id)}>
              <GrPauseFill />
            </button>
            <button onClick={() => this.props.abortWorker(worker.id)}>
              <GrFormClose />
            </button>
          </div>
        );
      }
    } else {
      return (
        <button
          className="wokerPanel_woker_deleteButton"
          onClick={() => this.props.deleteWorker(worker.id)}
        >
          <GrFormClose />
        </button>
      );
    }
  }

  /**
   * Function handling log lines
   * @param {sring} name Worker name
   * @param {number} lines log lines to show
   */
  getWorkerLog(name: string, lines: number) {
    this.props.onFetchWorkerLogs(name, lines);
    setTimeout(() => {
      const workerLogs = this.props.workerLogs;
      if (workerLogs !== false) {
        this.props.handleLogDialogContent(workerLogs);
      }
    }, 500);
  }

  /**
   * Handles the rendering of worker log dialog box
   * @param {string} worker worker name
   * @param {number} lines Lines to render
   */
  openLogModal(worker: string, lines: number) {
    this.getWorkerLog(worker, lines);
    this.props.openDialog();
  }

  render() {
    const { worker, logDialogContent } = this.props;
    const style = this.getPanelStyle();
    return (
      <div className="workerPanelContainer" data-test="workerPanelComponent">
        <div className="workerPanelContainer__headingContainer">
          <h1 className="workerPanelContainer__headingContainer__heading">
            {" "}
            {"Worker " + worker.id}
          </h1>
          <div className="workerPanelContainer__headingContainer__deleteButton pull-right">
            {this.getControlButtons()}
          </div>
        </div>

        <div className="workerPanelContainer__infoContainer">
          <p className="workerPanelContainer__infoContainer__pid">
            <strong>PID: </strong>
            {worker.worker}
          </p>
          {worker.start_time !== undefined ? (
            <p className="workerPanelContainer__infoContainer__startTime">
              <strong>Start Time: </strong>
              {worker.start_time + "(~"}{" "}
              <Moment date={worker.start_time} durationFromNow /> {")"}
            </p>
          ) : (
            <p className="workerPanelContainer__infoContainer__startTime">
              <strong>Start Time: </strong>N/A
            </p>
          )}
          {worker.work.length > 0 ? (
            <div className="workerPanelContainer__infoContainer__workerWorkContainer">
              <p>
                <strong>Target: </strong>
                <Link to={"/targets/" + worker.work[0]["id"]}>
                  {worker.work[0].target_url}
                </Link>
              </p>
              <p>
                <strong>Plugin: </strong>
                {worker.work[1].title}
              </p>
              <p>
                <strong>Type: </strong>
                {worker.work[1].type.replace("_", " ")}
              </p>
              <p>
                <strong>Group: </strong>
                {worker.work[1].group}
              </p>
            </div>
          ) : null}

          <button
            className="workerPanelContainer__infoContainer__showLogsButton"
            onClick={() => {
              this.displayLog();
            }}
          >
            Show logs
            {this.state.showLogs && (
              <div className="workerPanelContainer__infoContainer__showLogsButton__dropDownMenu">
                <span
                  onClick={() => {
                    this.hideLog();
                  }}
                >
                  None
                </span>
                <span onClick={() => this.openLogModal(worker.name, -1)}>
                  All
                </span>
                {[...Array(10)].map((_, i) => (
                  <span
                    key={i + 1}
                    // @ts-ignore
                    onClick={() => this.openLogModal(worker.name, i.toString())}
                  >
                    {i + 1}
                  </span>
                ))}
              </div>
            )}
          </button>
        </div>
      </div>
    );
  }
}
