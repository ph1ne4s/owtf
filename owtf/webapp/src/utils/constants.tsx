/*
 * OWTF UI configuration
 *
 * This file will contain global configuration.
 */

export const API_BASE_URL: string = "/api/v1/";
export const API_BASE_ISSUE_URL: string =
  "https://github.com/owtf/owtf/issues/";
