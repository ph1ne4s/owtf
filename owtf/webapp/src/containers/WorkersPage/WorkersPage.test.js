import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import "../../setupTests";
import { findByTestAtrr, checkProps } from "../../utils/testUtils";
import configureStore from "redux-mock-store";
import ConnectedWorkersPage, { WorkersPage } from "./index";
import WorkerPanel from "./WorkerPanel";
import { fromJS } from "immutable";

import {
  loadWorkers,
  workersLoaded,
  workersLoadingError,
  workerCreated,
  workerCreatingError,
  workerChanged,
  workerChangingError,
  workerDeleted,
  workerDeletingError
} from "./actions";

import {
  CHANGE_WORKER,
  CHANGE_WORKER_SUCCESS,
  CHANGE_WORKER_ERROR,
  LOAD_WORKERS,
  LOAD_WORKERS_SUCCESS,
  LOAD_WORKERS_ERROR,
  CREATE_WORKER,
  CREATE_WORKER_SUCCESS,
  CREATE_WORKER_ERROR,
  DELETE_WORKER,
  DELETE_WORKER_SUCCESS,
  DELETE_WORKER_ERROR
} from "./constants";

import { getWorkers, postWorker, patchWorker, deleteWorker } from "./saga";
import { runSaga } from "redux-saga";
import * as api from "./api";

import {
  workersLoadReducer,
  workerChangeReducer,
  workerCreateReducer,
  workerDeleteReducer
} from "./reducer";

const setUp = (initialState = {}) => {
  const mockStore = configureStore();
  const store = mockStore(initialState);
  const wrapper = shallow(<ConnectedWorkersPage store={store} />);
  return wrapper;
};

describe("Workers Page component", () => {
  describe("Testing dumb WorkersPage component", () => {
    let wrapper;
    let props;
    beforeEach(() => {
      props = {
        workers: [
          {
            busy: false,
            id: 4,
            name: "Worker-4",
            paused: false,
            work: [],
            worker: 29693
          }
        ],
        workerProgressLoading: false,
        workerProgressError: false,
        workerProgress: { left_count: 50, complete_count: 100 },
        workerLogs: false,
        fetchLoading: false,
        fetchError: false,
        changeError: false,
        changeLoading: false,
        deleteError: false,
        createError: false,
        onFetchWorkers: jest.fn(),
        onChangeWorker: jest.fn(),
        onDeleteWorker: jest.fn(),
        onCreateWorker: jest.fn(),
        onFetchWorkerProgress: jest.fn(),
        onFetchWorkerLogs: jest.fn()
      };
      wrapper = shallow(<WorkersPage {...props} />);
    });

    it("Should have correct prop-types", () => {
      const expectedProps = {
        workers: false,
        workerProgressLoading: false,
        workerProgressError: false,
        workerProgress: {},
        workerLogs: false,
        fetchLoading: false,
        fetchError: {},
        changeError: {},
        changeLoading: false,
        deleteError: false,
        createError: false,
        onFetchWorkers: () => {},
        onChangeWorker: () => {},
        onDeleteWorker: () => {},
        onCreateWorker: () => {},
        onFetchWorkerProgress: () => {},
        onFetchWorkerLogs: () => {}
      };
      const propsErr = checkProps(WorkersPage, expectedProps);
      expect(propsErr).toBeUndefined();
    });

    it("Should render without errors", () => {
      const component = findByTestAtrr(wrapper, "workerComponent");
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it("Should correctly render its sub-components", () => {
      const button = wrapper.find("button");
      const progressbar = wrapper.find("ProgressBar");
      const workerpanel = wrapper.find("WorkerPanel");

      expect(button.length).toBe(3);
      expect(progressbar.length).toBe(1);
      expect(workerpanel.length).toBe(props.workers.length);
      expect(button.at(0).props().children[1]).toEqual("Add worker !");
      expect(button.at(1).props().children[1]).toEqual("Pause All");
      expect(button.at(2).props().children[1]).toEqual("Resume All");

      wrapper.setProps({ workers: false });
      expect(wrapper.find("WorkerPanel").length).toBe(0);
    });

    it("Should call correct function after button click", () => {
      expect(props.onFetchWorkers.mock.calls.length).toBe(1);
      expect(props.onChangeWorker.mock.calls.length).toBe(0);
      expect(props.onCreateWorker.mock.calls.length).toBe(0);
      const addButton = wrapper.find("button").at(0);
      const pauseButton = wrapper.find("button").at(1);
      const resumeButton = wrapper.find("button").at(2);
      addButton.simulate("click");
      expect(props.onCreateWorker.mock.calls.length).toBe(1);
      pauseButton.simulate("click");
      expect(props.onChangeWorker.mock.calls.length).toBe(1);
      resumeButton.simulate("click");
      expect(props.onChangeWorker.mock.calls.length).toBe(2);
    });

    it("Should pass correct props to its child component", () => {
      const workerpanel = wrapper.find("WorkerPanel").at(0);
      expect(workerpanel.props().worker).toEqual(props.workers[0]);
    });
  });

  describe("Testing connected Workers Page component", () => {
    let wrapper;
    let initialState;
    beforeEach(() => {
      const workersLoad = {
        loading: false,
        error: false,
        workers: false
      };
      const workerCreate = {
        loading: true,
        error: false
      };
      const workerChange = {
        loading: true,
        error: false
      };
      const workerDelete = {
        loading: true,
        error: false
      };
      const workerProgressLoad = {
        loading: false,
        error: false,
        workerProgress: { left_count: 50, complete_count: 100 }
      };
      const workerLogsLoad = {
        loading: false,
        error: false,
        workerLogs: "Test worker logs"
      };
      const workers = {
        load: workersLoad,
        create: workerCreate,
        change: workerChange,
        delete: workerDelete,
        loadWorkerProgress: workerProgressLoad,
        loadWorkerLogs: workerLogsLoad
      };
      initialState = fromJS({
        workers
      });
      wrapper = setUp(initialState);
    });

    it("Props should match the initial state", () => {
      const workersProp = initialState
        .get("workers")
        .get("load")
        .get("workers");
      const fetchLoadingProp = initialState
        .get("workers")
        .get("load")
        .get("loading");
      const fetchErrorProp = initialState
        .get("workers")
        .get("load")
        .get("error");
      const changeErrorProp = initialState
        .get("workers")
        .get("change")
        .get("error");
      const deleteErrorProp = initialState
        .get("workers")
        .get("delete")
        .get("error");
      const createErrorProp = initialState
        .get("workers")
        .get("create")
        .get("error");
      const workerProgressProp = initialState
        .get("workers")
        .get("loadWorkerProgress")
        .get("workerProgress");
      const workerLogsProp = initialState
        .get("workers")
        .get("loadWorkerLogs")
        .get("workerLogs");

      expect(wrapper.props().workers).toEqual(workersProp);
      expect(wrapper.props().fetchLoading).toEqual(fetchLoadingProp);
      expect(wrapper.props().fetchError).toEqual(fetchErrorProp);
      expect(wrapper.props().changeError).toEqual(changeErrorProp);
      expect(wrapper.props().deleteError).toEqual(deleteErrorProp);
      expect(wrapper.props().createError).toEqual(createErrorProp);
      expect(wrapper.props().workerProgress).toEqual(workerProgressProp);
      expect(wrapper.props().workerLogs).toEqual(workerLogsProp);
    });
  });

  describe("Testing WorkerPanel component", () => {
    let wrapper;
    let props;
    beforeEach(() => {
      props = {
        worker: {
          busy: false,
          id: 4,
          name: "Worker-4",
          paused: false,
          work: [],
          worker: 29693
        },
        resumeWorker: jest.fn(),
        pauseWorker: jest.fn(),
        abortWorker: jest.fn(),
        deleteWorker: jest.fn(),
        handleLogDialogShow: jest.fn(),
        logDialogShow: false,
        onFetchWorkerLogs: jest.fn(),
        logDialogContent: "Test content",
        handleLogDialogContent: jest.fn()
      };
      wrapper = shallow(<WorkerPanel {...props} />);
    });

    it("Should have correct prop-types", () => {
      const expectedProps = {
        worker: {},
        resumeWorker: () => {},
        pauseWorker: () => {},
        abortWorker: () => {},
        deleteWorker: () => {},
        handleLogDialogShow: () => {},
        logDialogShow: true,
        onFetchWorkerLogs: () => {},
        logDialogContent: "",
        handleLogDialogContent: () => {}
      };
      const propsErr = checkProps(WorkerPanel, expectedProps);
      expect(propsErr).toBeUndefined();
    });

    it("Should render without errors", () => {
      const component = findByTestAtrr(wrapper, "workerPanelComponent");
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it("Should correctly render its sub-components", () => {
      const heading = wrapper.find("h1");

      expect(heading.length).toBe(1);
      expect(heading.props().children[1]).toEqual("Worker 4");

      const testWorker = {
        busy: true,
        id: 4,
        name: "Worker-4",
        paused: false,
        work: [],
        worker: 29693
      };
      wrapper.setProps({ worker: testWorker });
    });

    it("Should update state after logs button click", () => {
      expect(wrapper.instance().state.showLogs).toBe(false);
      const logbutton = wrapper.find(
        ".workerPanelContainer__infoContainer__showLogsButton"
      );
      logbutton.simulate("click");
      expect(wrapper.instance().state.showLogs).toBe(true);
    });

    it("Should correctly call function after icon button click", () => {
      expect(props.deleteWorker.mock.calls.length).toBe(0);
      const deleteiconbutton = wrapper.find(".wokerPanel_woker_deleteButton");
      deleteiconbutton.simulate("click");
      expect(props.deleteWorker.mock.calls.length).toBe(1);
    });
  });

  describe("Testing the sagas", () => {
    describe("Testing getWorkers saga", () => {
      it("Should load workers in case of success", async () => {
        // we push all dispatched actions to make assertions easier
        // and our tests less brittle
        const dispatchedActions = [];

        // we don't want to perform an actual api call in our tests
        // so we will mock the fetchAPI api with jest
        // this will mutate the dependency which we may reset if other tests
        // are dependent on it
        const mockedWorkers = {
          status: "success",
          data: [
            {
              busy: false,
              id: 4,
              name: "Worker-4",
              paused: false,
              work: [],
              worker: 29693
            }
          ]
        };
        api.getWorkersAPI = jest.fn(() =>
          jest.fn(() => Promise.resolve(mockedWorkers))
        );

        const fakeStore = {
          getState: () => ({ state: "test" }),
          dispatch: action => dispatchedActions.push(action)
        };

        // wait for saga to complete
        await runSaga(fakeStore, getWorkers).done;

        expect(api.getWorkersAPI.mock.calls.length).toBe(1);
        expect(dispatchedActions).toContainEqual(
          workersLoaded(mockedWorkers.data)
        );
      });

      it("Should handle workers load errors in case of failure", async () => {
        const dispatchedActions = [];

        // we simulate an error by rejecting the promise
        // then we assert if our saga dispatched the action(s) correctly
        const error = "API server is down";
        api.getWorkersAPI = jest.fn(() => jest.fn(() => Promise.reject(error)));

        const fakeStore = {
          getState: () => ({ state: "test" }),
          dispatch: action => dispatchedActions.push(action)
        };

        await runSaga(fakeStore, getWorkers).done;

        expect(api.getWorkersAPI.mock.calls.length).toBe(1);
        expect(dispatchedActions).toContainEqual(workersLoadingError(error));
      });
    });

    describe("Testing postWorker saga", () => {
      let fakeAction;
      beforeEach(() => {
        fakeAction = {
          type: CREATE_WORKER
        };
      });

      it("Should create a new worker and load workers in case of success", async () => {
        const dispatchedActions = [];
        api.postWorkerAPI = jest.fn(() => jest.fn(() => Promise.resolve()));

        const fakeStore = {
          getState: () => ({ state: "test" }),
          dispatch: action => dispatchedActions.push(action)
        };
        await runSaga(fakeStore, postWorker, fakeAction).done;

        expect(api.postWorkerAPI.mock.calls.length).toBe(1);
        expect(dispatchedActions).toContainEqual(workerCreated());
        expect(dispatchedActions).toContainEqual(loadWorkers());
      });

      it("Should handle worker create error in case of failure", async () => {
        const dispatchedActions = [];
        const error = "API server is down";
        api.postWorkerAPI = jest.fn(() => jest.fn(() => Promise.reject(error)));

        const fakeStore = {
          getState: () => ({ state: "test" }),
          dispatch: action => dispatchedActions.push(action)
        };

        await runSaga(fakeStore, postWorker, fakeAction).done;

        expect(api.postWorkerAPI.mock.calls.length).toBe(1);
        expect(dispatchedActions).toContainEqual(workerCreatingError(error));
      });
    });

    describe("Testing patchWorker saga", () => {
      let fakeAction;
      beforeEach(() => {
        fakeAction = {
          type: CHANGE_WORKER,
          worker_id: 1,
          action_type: "pause"
        };
      });

      it("Should change a worker and load workers in case of success", async () => {
        const dispatchedActions = [];
        api.patchWorkerAPI = jest.fn(() => jest.fn(() => Promise.resolve()));

        const fakeStore = {
          getState: () => ({ state: "test" }),
          dispatch: action => dispatchedActions.push(action)
        };
        await runSaga(fakeStore, patchWorker, fakeAction).done;

        expect(api.patchWorkerAPI.mock.calls.length).toBe(1);
        expect(dispatchedActions).toContainEqual(workerChanged());
        expect(dispatchedActions).toContainEqual(loadWorkers());
      });

      it("Should handle work change error in case of failure", async () => {
        const dispatchedActions = [];
        const error = "API server is down";
        api.patchWorkerAPI = jest.fn(() =>
          jest.fn(() => Promise.reject(error))
        );

        const fakeStore = {
          getState: () => ({ state: "test" }),
          dispatch: action => dispatchedActions.push(action)
        };

        await runSaga(fakeStore, patchWorker, fakeAction).done;

        expect(api.patchWorkerAPI.mock.calls.length).toBe(1);
        expect(dispatchedActions).toContainEqual(workerChangingError(error));
      });
    });

    describe("Testing deleteWorker saga", () => {
      let fakeAction;
      beforeEach(() => {
        fakeAction = {
          type: DELETE_WORKER,
          worker_id: 1
        };
      });

      it("Should delete a work and load workers in case of success", async () => {
        const dispatchedActions = [];
        api.deleteWorkerAPI = jest.fn(() => jest.fn(() => Promise.resolve()));

        const fakeStore = {
          getState: () => ({ state: "test" }),
          dispatch: action => dispatchedActions.push(action)
        };

        await runSaga(fakeStore, deleteWorker, fakeAction).done;

        expect(api.deleteWorkerAPI.mock.calls.length).toBe(1);
        expect(dispatchedActions).toContainEqual(workerDeleted());
        expect(dispatchedActions).toContainEqual(loadWorkers());
      });

      it("Should handle target worker deleting errors in case of failure", async () => {
        const dispatchedActions = [];
        const error = "API server is down";
        api.deleteWorkerAPI = jest.fn(() =>
          jest.fn(() => Promise.reject(error))
        );

        const fakeStore = {
          getState: () => ({ state: "test" }),
          dispatch: action => dispatchedActions.push(action)
        };

        await runSaga(fakeStore, deleteWorker, fakeAction).done;

        expect(api.deleteWorkerAPI.mock.calls.length).toBe(1);
        expect(dispatchedActions).toContainEqual(workerDeletingError(error));
      });
    });
  });

  describe("Testing reducers", () => {
    describe("Testing workersLoadReducer", () => {
      let initialState;
      beforeEach(() => {
        initialState = {
          loading: true,
          error: false,
          workers: false
        };
      });

      it("Should return the initial state", () => {
        const newState = workersLoadReducer(undefined, {});
        expect(newState.toJS()).toEqual(initialState);
      });

      it("Should handle LOAD_WORKERS", () => {
        const newState = workersLoadReducer(undefined, {
          type: LOAD_WORKERS
        });
        expect(newState.toJS()).toEqual({
          loading: true,
          error: false,
          workers: false
        });
      });

      it("Should handle LOAD_WORKERS_SUCCESS", () => {
        const workers = [
          {
            busy: false,
            id: 4,
            name: "Worker-4",
            paused: false,
            work: [],
            worker: 29693
          }
        ];
        const newState = workersLoadReducer(
          fromJS({
            loading: true,
            error: true,
            workers: false
          }),
          {
            type: LOAD_WORKERS_SUCCESS,
            workers
          }
        );
        expect(newState.toJS()).toEqual({
          loading: false,
          error: false,
          workers: workers
        });
      });

      it("Should handle LOAD_WORKERS_ERROR", () => {
        const newState = workersLoadReducer(
          fromJS({
            loading: true,
            error: true,
            workers: false
          }),
          {
            type: LOAD_WORKERS_ERROR,
            error: "Test workers loading error"
          }
        );
        expect(newState.toJS()).toEqual({
          loading: false,
          error: "Test workers loading error",
          workers: false
        });
      });
    });

    describe("Testing workerCreateReducer", () => {
      let initialState;
      beforeEach(() => {
        initialState = {
          loading: false,
          error: false
        };
      });

      it("Should return the initial state", () => {
        const newState = workerCreateReducer(undefined, {});
        expect(newState.toJS()).toEqual(initialState);
      });

      it("Should handle CREATE_WORKER", () => {
        const newState = workerCreateReducer(undefined, {
          type: CREATE_WORKER
        });
        expect(newState.toJS()).toEqual({ loading: true, error: false });
      });

      it("Should handle CREATE_WORKER_SUCCESS", () => {
        const newState = workerCreateReducer(
          fromJS({
            loading: true,
            error: true
          }),
          {
            type: CREATE_WORKER_SUCCESS
          }
        );
        expect(newState.toJS()).toEqual({
          loading: false,
          error: false
        });
      });

      it("Should handle CREATE_WORKER_ERROR", () => {
        const newState = workerCreateReducer(
          fromJS({
            loading: true,
            error: true
          }),
          {
            type: CREATE_WORKER_ERROR,
            error: "Test creating error"
          }
        );
        expect(newState.toJS()).toEqual({
          loading: false,
          error: "Test creating error"
        });
      });
    });

    describe("Testing workerChangeReducer", () => {
      let initialState;
      beforeEach(() => {
        initialState = {
          loading: false,
          error: false
        };
      });

      it("Should return the initial state", () => {
        const newState = workerChangeReducer(undefined, {});
        expect(newState.toJS()).toEqual(initialState);
      });

      it("Should handle CHANGE_WORKER", () => {
        const newState = workerChangeReducer(undefined, {
          type: CHANGE_WORKER,
          worker_id: 1,
          action_type: "pause"
        });
        expect(newState.toJS()).toEqual({ loading: true, error: false });
      });

      it("Should handle CHANGE_WORKER_SUCCESS", () => {
        const newState = workerChangeReducer(
          fromJS({
            loading: true,
            error: true
          }),
          {
            type: CHANGE_WORKER_SUCCESS
          }
        );
        expect(newState.toJS()).toEqual({
          loading: false,
          error: false
        });
      });

      it("Should handle CHANGE_WORKER_ERROR", () => {
        const newState = workerChangeReducer(
          fromJS({
            loading: true,
            error: true
          }),
          {
            type: CHANGE_WORKER_ERROR,
            error: "Test changing error"
          }
        );
        expect(newState.toJS()).toEqual({
          loading: false,
          error: "Test changing error"
        });
      });
    });

    describe("Testing workerDeleteReducer", () => {
      let initialState;
      beforeEach(() => {
        initialState = {
          loading: false,
          error: false
        };
      });

      it("Should return the initial state", () => {
        const newState = workerDeleteReducer(undefined, {});
        expect(newState.toJS()).toEqual(initialState);
      });

      it("Should handle DELETE_WORKER", () => {
        const newState = workerDeleteReducer(undefined, {
          type: DELETE_WORKER,
          worker_id: 1
        });
        expect(newState.toJS()).toEqual({ loading: true, error: false });
      });

      it("Should handle DELETE_WORKER_SUCCESS", () => {
        const newState = workerDeleteReducer(
          fromJS({
            loading: true,
            error: true
          }),
          {
            type: DELETE_WORKER_SUCCESS
          }
        );
        expect(newState.toJS()).toEqual({
          loading: false,
          error: false
        });
      });

      it("Should handle DELETE_WORKER_ERROR", () => {
        const newState = workerDeleteReducer(
          fromJS({
            loading: true,
            error: true
          }),
          {
            type: DELETE_WORKER_ERROR,
            error: "Test deleting error"
          }
        );
        expect(newState.toJS()).toEqual({
          loading: false,
          error: "Test deleting error"
        });
      });
    });
  });
});
