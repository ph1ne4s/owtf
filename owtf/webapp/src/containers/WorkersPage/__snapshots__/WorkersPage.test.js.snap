// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workers Page component Testing WorkerPanel component Should render without errors 1`] = `
<div
  className="workerPanelContainer"
  data-test="workerPanelComponent"
>
  <div
    className="workerPanelContainer__headingContainer"
  >
    <h1
      className="workerPanelContainer__headingContainer__heading"
    >
       
      Worker 4
    </h1>
    <div
      className="workerPanelContainer__headingContainer__deleteButton pull-right"
    >
      <button
        className="wokerPanel_woker_deleteButton"
        onClick={[Function]}
      >
        <GrFormClose />
      </button>
    </div>
  </div>
  <div
    className="workerPanelContainer__infoContainer"
  >
    <p
      className="workerPanelContainer__infoContainer__pid"
    >
      <strong>
        PID: 
      </strong>
      29693
    </p>
    <p
      className="workerPanelContainer__infoContainer__startTime"
    >
      <strong>
        Start Time: 
      </strong>
      N/A
    </p>
    <button
      className="workerPanelContainer__infoContainer__showLogsButton"
      onClick={[Function]}
    >
      Show logs
    </button>
  </div>
</div>
`;

exports[`Workers Page component Testing dumb WorkersPage component Should render without errors 1`] = `
<div
  className="workersPageContainer"
  data-test="workerComponent"
>
  <div
    className="workersPageContainer__buttonsContainer"
  >
    <button
      className="workersPageContainer__buttonsContainer__addWorkerButton"
      onClick={[Function]}
    >
      <BsPlusLg />
      Add worker !
    </button>
    <button
      className="workersPageContainer__buttonsContainer__pauseAllButton"
      onClick={[Function]}
    >
      <GiPauseButton />
      Pause All
    </button>
    <button
      className="workersPageContainer__buttonsContainer__resumeAllButton"
      onClick={[Function]}
    >
      <BsPlayFill />
      Resume All
    </button>
  </div>
  <h2
    className="workersPageContainer__heading"
    title="updates in 13s"
  >
    Total scan progress
  </h2>
  <div
    className="workersPageContainer__progressBarContainer"
  >
    <ProgressBar
      active={true}
      bsClass="progress-bar"
      bsStyle="success"
      isChild={false}
      max={100}
      min={0}
      now={66.66666666666666}
      srOnly={false}
      striped={true}
    />
  </div>
  <div
    className="workersPageContainer__workerPanelContainer"
  >
    <WorkerPanel
      abortWorker={[Function]}
      deleteWorker={[Function]}
      handleLogDialogContent={[Function]}
      handleLogDialogShow={[Function]}
      key="4"
      logDialogContent="Nothing to show here!"
      onFetchWorkerLogs={[MockFunction]}
      openDialog={[Function]}
      pauseWorker={[Function]}
      resumeWorker={[Function]}
      worker={
        Object {
          "busy": false,
          "id": 4,
          "name": "Worker-4",
          "paused": false,
          "work": Array [],
          "worker": 29693,
        }
      }
      workerLogs={false}
    />
  </div>
  <div
    className="dialogWrapper"
  >
    <Dialog
      isDialogOpened={false}
      onClose={[Function]}
      title="Worker Log"
    >
      <pre
        style={
          Object {
            "maxHeight": "80vh",
            "overflow": "scroll",
          }
        }
      >
        Nothing to show here!
      </pre>
    </Dialog>
  </div>
</div>
`;
